File: src/main/java/com/senderman/lastkatkabot/command/user/MarryMe.java
package com.senderman.lastkatkabot.command.user;

import com.annimon.tgbotsmodule.api.methods.Methods;
import com.annimon.tgbotsmodule.services.CommonAbsSender;
import com.senderman.lastkatkabot.ApiRequests;
import com.senderman.lastkatkabot.callback.Callbacks;
import com.senderman.lastkatkabot.command.CommandExecutor;
import com.senderman.lastkatkabot.dbservice.UserStatsService;
import com.senderman.lastkatkabot.model.MarriageRequest;
import com.senderman.lastkatkabot.repository.MarriageRequestRepository;
import com.senderman.lastkatkabot.util.Html;
import com.senderman.lastkatkabot.util.callback.ButtonBuilder;
import com.senderman.lastkatkabot.util.callback.MarkupBuilder;
import org.springframework.stereotype.Component;
import org.telegram.telegrambots.meta.api.objects.Message;

import java.util.Objects;

@Component
public class MarryMe implements CommandExecutor {

    private final CommonAbsSender telegram;
    private final UserStatsService users;
    private final MarriageRequestRepository marriages;

    public MarryMe(CommonAbsSender telegram, UserStatsService users, MarriageRequestRepository marriages) {
        this.telegram = telegram;
        this.users = users;
        this.marriages = marriages;
    }

    @Override
    public String getTrigger() {
        return "/marryme";
    }

    @Override
    public String getDescription() {
        return "(reply) пожениться на ком-нибудь";
    }

    @Override
    public void execute(Message message) {
        var chatId = message.getChatId();
        if (message.isUserMessage() || !message.isReply()) {
            ApiRequests.answerMessage(message, "Для использования команды необходимо ответить ей на чье-нибудь сообщение!")
                    .callAsync(telegram);
            return;
        }

        var proposerId = message.getFrom().getId();
        var proposeeId = message.getReplyToMessage().getFrom().getId();

        if (proposerId.equals(proposeeId)) {
            Methods.sendMessage(chatId, "На самом себе нельзя жениться!").callAsync(telegram);
            return;
        }

        var proposerStats = users.findById(proposerId);

        if (Objects.equals(proposerStats.getLoverId(), proposeeId)) {
            ApiRequests.answerMessage(message, "Вы уже в браке с этим пользователем!").callAsync(telegram);
            return;
        }

        if (proposerStats.hasLover()) {
            ApiRequests.answerMessage(message, "Вы что, хотите изменить своей половинке?!").callAsync(telegram);
            return;
        }
        var proposeeStats = users.findById(proposeeId);

        if (proposeeStats.hasLover()) {
            ApiRequests.answerMessage(message, "У этого пользователя уже есть своя вторая половинка!").callAsync(telegram);
            return;
        }

        var proposerLink = Html.getUserLink(message.getFrom());
        var text = "Пользователь " + proposerLink + " предлагает вам предлагает вам руку, сердце и шавуху. Вы согласны?";

        var request = new MarriageRequest();
        request.setId(marriages.findFirstByOrderByIdDesc().map(r -> r.getId() + 1).orElse(1));
        request.setProposerId(proposerId);
        request.setProposerName(proposerLink);
        request.setProposeeId(proposeeId);
        request.setProposeeName(Html.getUserLink(message.getReplyToMessage().getFrom()));
        request.setRequestDate(message.getDate());
        marriages.save(request);

        var markup = new MarkupBuilder()
                .addButton(ButtonBuilder.callbackButton()
                        .text("Принять")
                        .payload(Callbacks.MARRIAGE + " " + request.getId() + " accept"))
                .addButton(ButtonBuilder.callbackButton()
                        .text("Отказаться")
                        .payload(Callbacks.MARRIAGE + " " + request.getId() + " decline"))
                .build();

        Methods.sendMessage()
                .setChatId(chatId)
                .setText(text)
                .setReplyToMessageId(message.getReplyToMessage().getMessageId())
                .setReplyMarkup(markup)
                .callAsync(telegram);
    }

}

File: src/main/java/com/senderman/lastkatkabot/usercommands/MarryMe.kt
package com.senderman.lastkatkabot.usercommands

import com.annimon.tgbotsmodule.api.methods.Methods
import com.senderman.CommandExecutor
import com.senderman.TgUser
import com.senderman.lastkatkabot.LastkatkaBot
import com.senderman.lastkatkabot.LastkatkaBotHandler
import com.senderman.lastkatkabot.Services
import org.telegram.telegrambots.meta.api.objects.Message
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton

class MarryMe constructor(val handler: LastkatkaBotHandler): CommandExecutor {
    override val command: String
        get() ="/marryme"
    override val desc: String
        get() = "(reply) пожениться на ком-нибудь"

    override fun execute(message: Message) {
        val marryById = message.text.trim().matches(Regex("/marryme\\s+\\d+"))
        val chatId = message.chatId
        val userId = message.from.id
        val text: String
        val loverId: Int
        if (Services.db.getLover(userId) != 0) {
            handler.sendMessage(chatId, "Всмысле? Вы что, хотите изменить своей второй половинке?!")
            return
        }

        if (!marryById) {
            if (!message.isReply
                    || message.from.id == message.replyToMessage.from.id || message.replyToMessage.from.bot) return
            loverId = message.replyToMessage.from.id
            val user = TgUser(Methods.getChatMember(chatId, userId).call(handler).user)
            text = "Пользователь " + user.link + " предлагает вам руку, сердце и шавуху. Вы согласны?"

        } else {
            if (message.isUserMessage) return
            loverId = try {
                message.text.split(" ")[1].toInt()
            } catch (e: NumberFormatException) {
                handler.sendMessage(chatId, "Неверный формат!")
                return
            }
            val user = TgUser(Methods.getChatMember(chatId, userId).call(handler).user)
            val lover = TgUser(Methods.getChatMember(chatId, loverId).call(handler).user)
            text = "${lover.link}, пользователь ${user.link} предлагает вам руку, сердце и шавуху. Вы согласны?"
        }
        if (Services.db.getLover(loverId) != 0) {
            handler.sendMessage(chatId, "У этого пользователя уже есть своя вторая половинка!")
            return
        }
        val markup = InlineKeyboardMarkup()
        markup.keyboard = listOf(listOf(
                InlineKeyboardButton().apply {
                    this.text = "Принять"
                    callbackData = LastkatkaBot.CALLBACK_ACCEPT_MARRIAGE + "$userId $loverId"

                },
                InlineKeyboardButton().apply {
                    this.text = "Отказаться"
                    callbackData = LastkatkaBot.CALLBACK_DENY_MARRIAGE + "$userId $loverId"
                }
        ))
        val sm = Methods.sendMessage()
                .setChatId(chatId)
                .setText(text)
                .setReplyMarkup(markup)
        if (!marryById) {
            sm.replyToMessageId = message.replyToMessage.messageId
        }
        handler.sendMessage(sm)
    }
}